File: genesis/src/main/java/net/spartanb312/genesis/java/builder/FieldBuilder.java
package net.spartanb312.genesis.java.builder;

import net.spartanb312.genesis.java.Utils;
import org.objectweb.asm.tree.AnnotationNode;
import org.objectweb.asm.tree.FieldNode;

public abstract class FieldBuilder {

    public final FieldNode fieldNode;

    public FieldBuilder(
            int access,
            String name,
            String descriptor,
            String signature,
            Object value
    ) {
        fieldNode = new FieldNode(access, name, descriptor, signature, value);
    }

    public FieldBuilder(
            int access,
            String name,
            String descriptor,
            Object value
    ) {
        fieldNode = new FieldNode(access, name, descriptor, null, value);
    }

    public FieldBuilder(
            int access,
            String name,
            String descriptor
    ) {
        fieldNode = new FieldNode(access, name, descriptor, null, null);
    }

    public abstract void assemble();

    public FieldNode toFieldNode() {
        assemble();
        return fieldNode;
    }

    public AnnotationNode addAnnotation(AnnotationNode annotation) {
        fieldNode.visibleAnnotations = Utils.add(fieldNode.visibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationNode annotation, boolean visible) {
        if (visible) fieldNode.visibleAnnotations = Utils.add(fieldNode.visibleAnnotations, annotation);
        else fieldNode.invisibleAnnotations = Utils.add(fieldNode.invisibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationBuilder builder) {
        AnnotationNode annotation = builder.toAnnotationNode();
        fieldNode.visibleAnnotations = Utils.add(fieldNode.visibleAnnotations, annotation);
        return annotation;
    }

    public AnnotationNode addAnnotation(AnnotationBuilder builder, boolean visible) {
        AnnotationNode annotation = builder.toAnnotationNode();
        if (visible) fieldNode.visibleAnnotations = Utils.add(fieldNode.visibleAnnotations, annotation);
        else fieldNode.invisibleAnnotations = Utils.add(fieldNode.invisibleAnnotations, annotation);
        return annotation;
    }

}

File: genesis/src/main/kotlin/net/spartanb312/genesis/FieldBuilder.kt
package net.spartanb312.genesis

import net.spartanb312.genesis.extensions.BuilderDSL
import net.spartanb312.genesis.extensions.Modifiers
import net.spartanb312.genesis.extensions.NodeDSL
import org.objectweb.asm.tree.AnnotationNode
import org.objectweb.asm.tree.FieldNode

@JvmInline
value class FieldBuilder(val fieldNode: FieldNode) {

    @BuilderDSL
    operator fun AnnotationNode.unaryPlus() = fieldNode.visibleAnnotations.add(this)

    @BuilderDSL
    operator fun InvisibleAnnotationNode.unaryPlus() = fieldNode.invisibleAnnotations.add(this)

}

@NodeDSL
inline fun FieldNode.modify(block: FieldBuilder.() -> Unit): FieldNode {
    FieldBuilder(this).block()
    return this
}

@NodeDSL
fun field(
    access: Modifiers,
    name: String,
    desc: String,
    signature: String? = null,
    value: Any? = null
): FieldNode = FieldNode(access.modifier, name, desc, signature, value)

@NodeDSL
fun field(
    access: Int,
    name: String,
    desc: String,
    signature: String? = null,
    value: Any? = null
): FieldNode = FieldNode(access, name, desc, signature, value)


@NodeDSL
inline fun field(
    access: Modifiers,
    name: String,
    desc: String,
    signature: String? = null,
    value: Any? = null,
    block: FieldBuilder.() -> Unit
): FieldNode = FieldNode(access.modifier, name, desc, signature, value).modify(block)

@NodeDSL
inline fun field(
    access: Int,
    name: String,
    desc: String,
    signature: String? = null,
    value: Any? = null,
    block: FieldBuilder.() -> Unit
): FieldNode = FieldNode(access, name, desc, signature, value).modify(block)
